appModuleHandler.AppModule.appArchitecture: detect AMD64 apps in Windows 11 on ARM via kernel32.dll::GetProcessInformation

josephsl opened the pull request:
### Link to issue number:
Closes #14403 
Stems from #14397 

### Summary of the issue:
In Windows 11 on ARM, AMD64 apps are seen as ARM64 apps.

### Description of user facing changes
Entering "focus.appModule.appArchitecture" in Python Console will say "AMD64" if focused on AMD64 application.

### Description of development approach
In app module handler:

1. Defined a process information structure with three attributes: process machine (unsigned short), res0 (double word), machine attributes (double word.
2. Added 0xAA64 to represent ARM64 applications in app architecture property.
3. Refined app architecture property in AppModule class to call kernel32.dll::GetProcessInformation on Windows 11 21H2 (build 22000) or later, with WoW64 detection reserved for Windows 10 and earlier.
4. If GetProcessInformation fails (returns 0), report "unknown" as app architecture to prepare to handle cases where an app (perhaps hosted or an ARM64EC) may report wrong architecture information.
5. Added type information to app architecture property (return type: str). This should inspire efforts to add type information to app module handler odule and classes (later).

### Testing strategy:
Manual testing (ARM64 machine is ideal but also works on x64 machines):

1. If not done so, upgrade to Windows 11 on a supported ARM-based computer (Surface Pro X, for example).
2. Install and run an application known to be designed for AMD64 such as Notepad++ or Slack.
3. While focused on the app, press Control+NVDA+Z to open Python Console and type "focus.appModule.appArchitecture" and observe the value (expected: AMD64).

### Known issues with pull request:
None, although it might be possible that some x64 apps may continue to be seen as ARM64, in which case the development approach may need to be refined.

### Change log entries:
Bug fixes (and changes for developers):

In Windows 11 on ARM, x64 apps are no longer identified as ARM64 applications. (#14403)

### Code Review Checklist:

<!--
This checklist is a reminder of things commonly forgotten in a new PR.
Authors, please do a self-review of this pull-request.
Check items to confirm you have thought about the relevance of the item.
Where items are missing (eg unit / system tests), please explain in the PR.
To check an item `- [ ]` becomes `- [x]`, note spacing.
You can also check the checkboxes after the PR is created.
A detailed explanation of this checklist is available here:
https://github.com/nvaccess/nvda/blob/master/devDocs/githubPullRequestTemplateExplanationAndExamples.md#code-review-checklist
-->

- [x] Pull Request description:
  - description is up to date
  - change log entries
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] API is compatible with existing add-ons.
- [x] Documentation:
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] Security precautions taken.

josephsl committed:
appModuleHandler.AppModule: define process information structure for use when detecting various ap architectures in Windows 11. Re #14403.

In Windows 11 on ARM, 64-bit x86 (AMD64) apps are identified as ARM64 when it should say AMD64, traced to a flaw in AppModule.appArchitecture property that relied on WoW64 detection. As part of resolving this, define a process machine info structure to be used in the refined app architecture property getter.

josephsl committed:
AppModule.appArchitecture: call kernel32::GetProcessInformation to obtain app architecture in Windows 11. Re #14403.

Rather than relying on WoW64 detection, call kernel32.dll::GetProcessInformation in Windows 11 (21H2 build 22000 and later) to obtain app architecture. This uses a constant defined in process information class enumeration to obtain machine info (9). The WoW64 method is used if this is Windows 10 or earlier.

josephsl committed:
AppModule.appArchitecture: recognize 0xAA64 as ARM64 machine type. Re #14403

josephsl committed:
AppModule.appArchitecture: winKernel -> ctypes.windll for consistency. Re #14403

josephsl committed:
AppModule.appArchitecture: add type information (return type: str). Re #14403

rperez030 commented:
Tested using a Windows on ARM VM. X86, AMD64 and ARM64 where reported correctly for Slack (X86), Remote Incident Manager (AMD64) and Visual Studio code (ARM64), however, when tested with 1Password 8 which should have been reported as AMD64, NVDA generated the following error:

Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "baseObject.pyc", line 26, in __get__
  File "appModuleHandler.pyc", line 710, in _get_appArchitecture
KeyError: 0

NVDA 2022.3.2 incorrectly reports ARM64 for the same app.

josephsl commented:
Hi, sounds like it is saying “machine unknown” for this particular app. We might as well say “ARM64” for this case unless this is truly an AMD64 app and/or is an EC app. CC @MichaelDCurran both for your comment and a review please.

rperez030 commented:
I believe it is an AMD64, or maybe EC. I'm going by their website where they indicate that it requires Windows 11 64-bit or Windows 10 64-bit

leonardder commented:
I'm indeed pretty sure that 1Password is AMD64 only.

michaelDCurran commented on source/appModuleHandler.py:
```diff
@@ -682,28 +690,40 @@ def _get_appArchitecture(self):
+			processMachineInfo = _PROCESS_MACHINE_INFORMATION()
+			# Constant comes from PROCESS_INFORMATION_CLASS enumeration.
+			ProcessMachineTypeInfo = 9
+			ctypes.windll.kernel32.GetProcessInformation(
```
Definitely should check the return value here. If false then we should report as unknown.
My hypothesis is that for 1Password  this function fails due to process handle permissions.
However, if this is not the case, then at very least, using `get` or catching KeyError on the arch dictionary will obviously be necessary.

josephsl committed:
Merge branch 'master' into i14403win11GetProcessInformation

josephsl committed:
appModuleHandler.AppModule: handle kernel32.dll::GetProcessInformation returning 0 (error). Re #14403.

Comment from Mick Curran (NV Access): GetProcessInformation may say '0' (error) sometimes, therefore treat the app in question has targeting 'unknown' architecture.

michaelDCurran commented on source/appModuleHandler.py:
```diff
@@ -684,28 +692,43 @@ def _get_appArchitecture(self):
+				ctypes.byref(processMachineInfo),
+				ctypes.sizeof(_PROCESS_MACHINE_INFORMATION)
+			):
+				self.appArchitecture = "unknown"
```
```suggestion
				log.debugWarning("Could not query process architecture")
				self.appArchitecture = "unknown"
```

michaelDCurran commented on source/appModuleHandler.py:
```diff
@@ -684,28 +692,43 @@ def _get_appArchitecture(self):
-		except AttributeError:
-			# Windows 10 Version 1507 (build 10240) and earlier.
-			self.appArchitecture = "AMD64" if self.is64BitProcess else "x86"
+				self.appArchitecture = archValues2ArchNames.get(processMachineInfo.ProcessMachine, "unknown")
```
Add debug warnings for both situations. 
```suggestion
				try:
					self.appArchitecture = archValues2ArchNames[processMachineInfo.ProcessMachine]
				except KeyError:
					log.debugWarning(f"Unknown process architecture 0X{processMachineInfo.ProcessMachine:X}")
					self.appArchitecture = "unknown"
```

michaelDCurran commented:
Having tested this on ARM64 now:
The appArchitecture of 1Password is failing because getProcessInformation returns 0, because the appModule's processHandle is actually 0. I.e. NVDA was unable to OpenProcess on 1Password.
However, the processHandle property of any NVDAObject from 1Password is valid. Oleacc's GetProcessHandleFromHwnd is working.
Therefore, it might be an idea to update an appModule's processHandle property if it is 0, with an NvDAObject's valid one.
I would suggest doing this in `appModuleHandler.GetAppModuleForNVDAObject`.
Btw, 1Password is in deed AMD64 on ARM64.

michaelDCurran commented:
It is also worth noting that in the 1password case, the appModule is also staying alive for ever, as passing a processHandle of 0 to WaitForSingleObject is returning a True value. (WAIT_FAIL). We should consider improving the isAlive property to only return True for WAIT_TIMEOUT, False for WAIT_0, and anything else should be logged as n error and probably return False.

josephsl commented:
Hi,

I'll incorporate Mick's suggestion this week after merging latest master commit on AMD64 support on Windows 11 ARM64.

Thanks.

josephsl committed:
Merge branch 'master' into i14403win11GetProcessInformation

josephsl committed:
App module handler: refine get app module from NVDA object function to set process handle to a valid one if it is 0. Re #14403.

Comment from ick Curran (NV Access): GetProcessInformation fials for some apps because process handle is 0 but the handle reported by NVDA object is valid. Therefore, after obtaining app module based on object process Id, see if process handle for the app is 0, and if so, replace process handle with that of the one reported by NVDA object. Also added type information for getAppModuleFromNVDAObject function.

josephsl committed:
App module handler: catch atribute error of obj.processHandle is undefined. Re #14403.

When running unit tests, a placeholder NVDA object is created with process handle undefined. Therefore catch attribute error and move on.

josephsl commented:
Hi,

Perhaps a follow-up work: when running unit tests, process handle is undefined for placeholder NVDA object. For now NVDA will ignore it via attribute error.

Thanks.

josephsl commented:
Hi,

Pull request build: https://ci.appveyor.com/api/buildjobs/x6a2kw2muycblho7/artifacts/output%2Fnvda_snapshot_pr14404-27340%2C5a08f6a6.exe

While running this build, see if 1password is recognized as an AMD64 application under ARM64 (open NVDA Pyhton Console, type focus.appModule.appArchitecture).

Thanks.

seanbudd committed:
Merge remote-tracking branch 'origin/master' into i14403win11GetProcessInformation

seanbudd committed:
update changes

seanbudd committed:
fix commit
